src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl

```diff
@@ -16,15 +16,73 @@ end
   return all(x->radical_membership(x, defining_ideal(P)), gens(ambient_coordinate_ring(P)))
 end
 
+@doc raw"""
+    function is_smooth(P::AbsProjectiveScheme; algorithm=:default) -> Bool
 
-# TODO: Jacobi criterion
-function is_smooth(P::AbsProjectiveScheme; algorithm=:covered)
+Check whether the scheme `X` is smooth.
+
+# Algorithms
+
+There are three possible algorithms for checking smoothness, determined
+by the value of the keyword argument `algorithm`:
+  * `:covered` - converts to a covered scheme,
+  * `:jacobi` - uses the Jacobian criterion for projective varieties,
+    see Exercise 4.2.10 of [Liu06](@cite),
+  * `:affine_cone` - checks that the affine cone is smooth outside the origin.
+
+The `:jacobi` algorithm first checks that the scheme is integral,
+which can be expensive.
+
+By default, if the base ring is a field and the scheme has already been
+computed to be integral, then the `:jacobi` algorithm is used.
+Otherwise, the `:affine_cone` algorithm is used.
+
+# Examples
+```jldoctest
+julia> A, (x, y, z) = grade(QQ["x", "y", "z"][1]);
+
+julia> B, _ = quo(A, ideal(A, [x^2 + y^2]));
+
+julia> C = projective_scheme(B)
+Projective scheme
+  over rational field
+defined by ideal(x^2 + y^2)
+
+julia> is_smooth(C)
+false
+
+julia> is_smooth(C; algorithm=:jacobi)
+false
+```
+"""
+function is_smooth(P::AbsProjectiveScheme; algorithm=:default)
   get_attribute!(P, :is_smooth) do
+    if algorithm == :default
+      if base_ring(P) isa Field && has_attribute(P, :is_integral) && is_integral(P)
+        algorithm = :jacobi
+      else
+        algorithm = :affine_cone
+      end
+    end
     if algorithm == :covered
       return is_smooth(covered_scheme(P))
-    end
-    if algorithm == :jacobi
-      error("jacobi criterion not implemented")
+    elseif algorithm == :affine_cone
+      aff = affine_cone(P)[1]
+      origin = subscheme(aff, coordinates(aff))
+      return singular_locus_reduced(aff)[1] == origin
```

Comment:
This equality check uses a radical computation and an equality check. Could you check whether some variant using saturation or radical membership might be more efficient?

Reply:
I benchmarked three approaches on the example `castelnuovo()`:

* `singular_locus_reduced`,
* saturation,
* radical membership.

They all seemed roughly the same speed.